File: frontend/src/app/services/token.service.ts
import { Injectable } from "@angular/core";
import { JwtHelperService } from "@auth0/angular-jwt"

@Injectable({
    providedIn: 'root',
})

export class TokenService {
    private readonly TOKEN_KEY = 'access_token';
    private jwtHelperService = new JwtHelperService();

    constructor() { }

    // getter and setter
    getToken(): string | null {
        return localStorage.getItem(this.TOKEN_KEY) || sessionStorage.getItem(this.TOKEN_KEY);
    }

    setToken(token: string, rememberMe: boolean = false): void {
        if (rememberMe) {
            localStorage.setItem(this.TOKEN_KEY, token);
        }
        else {
            sessionStorage.setItem(this.TOKEN_KEY, token);
        }
    }

    removeToken(): void {
        localStorage.removeItem(this.TOKEN_KEY);
        sessionStorage.removeItem(this.TOKEN_KEY);
    }

    getUserId(): number {
        debugger
        const token = this.getToken();
        if (!token) return 1;

        let accountObject = this.jwtHelperService.decodeToken(this.getToken() ?? '');
        if (!accountObject || typeof accountObject !== 'object') return 0;

        return 'userId' in accountObject ? parseInt(accountObject['userId']) : 0;
    }

    isTokenExpired(): boolean {
        debugger
        if (this.getToken() == null) {
            return false;
        }
        return this.jwtHelperService.isTokenExpired(this.getToken());
    }
}

File: android/lib/services/token_service.dart
import 'package:jwt_decoder/jwt_decoder.dart';
import 'package:shared_preferences/shared_preferences.dart';

class TokenService {
  static const String _tokenKey = 'access_token';
  static String? _memoryToken;

  static Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();

    return _memoryToken ?? prefs.getString(_tokenKey);
  }

  static Future<void> setToken(String token, {
    bool rememberMe = false,
  }) async {
    final prefs = await SharedPreferences.getInstance();

    if (rememberMe) {
      await prefs.setString(_tokenKey, token);
      _memoryToken = null;
    } else {
      _memoryToken = token;
    }
  }

  static Future<void> removeToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    _memoryToken = null;
  }

  static Future<int> getUserId() async {
    final token = await getToken();
    if (token == null) return 0;

    final decoded = JwtDecoder.decode(token);

    if (!decoded.containsKey('userId')) return 0;

    return int.tryParse(decoded['userId'].toString()) ?? 0;
  }

  static Future<bool> isTokenExpired() async {
    final token = await getToken();
    if (token == null) return true;

    return JwtDecoder.isExpired(token);
  }
}
